coderd/tailnet.go

```diff
@@ -0,0 +1,339 @@
+	"github.com/coder/coder/codersdk"
+	"github.com/coder/coder/site"
+	"github.com/coder/coder/tailnet"
+)
+
+var tailnetTransport *http.Transport
+
+func init() {
+	var valid bool
+	tailnetTransport, valid = http.DefaultTransport.(*http.Transport)
+	if !valid {
+		panic("dev error: default transport is the wrong type")
+	}
+}
+
+// NewServerTailnet creates a new tailnet intended for use by coderd. It
+// automatically falls back to wsconncache if a legacy agent is encountered.
+func NewServerTailnet(
+	ctx context.Context,
+	logger slog.Logger,
+	derpServer *derp.Server,
+	derpMap *tailcfg.DERPMap,
+	coord *atomic.Pointer[tailnet.Coordinator],
+	cache *wsconncache.Cache,
+) (*ServerTailnet, error) {
+	logger = logger.Named("servertailnet")
+	conn, err := tailnet.NewConn(&tailnet.Options{
+		Addresses: []netip.Prefix{netip.PrefixFrom(tailnet.IP(), 128)},
+		DERPMap:   derpMap,
+		Logger:    logger,
+	})
+	if err != nil {
+		return nil, xerrors.Errorf("create tailnet conn: %w", err)
+	}
+
+	serverCtx, cancel := context.WithCancel(ctx)
+	tn := &ServerTailnet{
+		ctx:        serverCtx,
+		cancel:     cancel,
+		logger:     logger,
+		conn:       conn,
+		coord:      coord,
+		cache:      cache,
+		agentNodes: map[uuid.UUID]*tailnetNode{},
+		transport:  tailnetTransport.Clone(),
+	}
+	tn.transport.DialContext = tn.dialContext
+	tn.transport.MaxIdleConnsPerHost = 10
+	tn.transport.MaxIdleConns = 0
+	agentConn := (*coord.Load()).ServeMultiAgent(uuid.New())
+	tn.agentConn.Store(&agentConn)
+
+	err = tn.getAgentConn().UpdateSelf(conn.Node())
+	if err != nil {
+		tn.logger.Warn(context.Background(), "server tailnet update self", slog.Error(err))
+	}
+	conn.SetNodeCallback(func(node *tailnet.Node) {
+		err := tn.getAgentConn().UpdateSelf(node)
+		if err != nil {
+			tn.logger.Warn(context.Background(), "broadcast server node to agents", slog.Error(err))
+		}
+	})
+
+	// This is set to allow local DERP traffic to be proxied through memory
+	// instead of needing to hit the external access URL. Don't use the ctx
+	// given in this callback, it's only valid while connecting.
+	conn.SetDERPRegionDialer(func(_ context.Context, region *tailcfg.DERPRegion) net.Conn {
+		if !region.EmbeddedRelay {
+			return nil
+		}
+		left, right := net.Pipe()
+		go func() {
+			defer left.Close()
+			defer right.Close()
+			brw := bufio.NewReadWriter(bufio.NewReader(right), bufio.NewWriter(right))
+			derpServer.Accept(ctx, right, brw, "internal")
+		}()
+		return left
+	})
+
+	go tn.watchAgentUpdates()
+	go tn.expireOldAgents()
+	return tn, nil
+}
+
+func (s *ServerTailnet) expireOldAgents() {
+	const (
+		tick   = 5 * time.Minute
+		cutoff = 30 * time.Minute
+	)
+
+	ticker := time.NewTicker(tick)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case <-ticker.C:
+		}
+
+		s.nodesMu.Lock()
+		agentConn := s.getAgentConn()
+		for agentID, node := range s.agentNodes {
+			if time.Since(node.lastConnection) > cutoff {
```

Comment:
This measures the last time we _started_ a connection, not the last time the connection was used.  If we proxy a long-lived connection like `ReconnectingPTY` or a devURL websocket, it could easily be in use for greater than 30 minutes.

We might need some ref-counting to keep track of the connections to each agent, so that we expire them when they are no longer used.